### DISCUSSION TITLE
Getting "Ill-formed input" in ostree sign

### DISCUSSION THREAD

**Discussion by ganjiii3** [ostreedev/ostree#3345] [Q&A] [answered]
Hi!
I wanted to try composefs with the verification of the userspace by `ostree-prepare-root`, however I am not able to pass the ed25519 key to `ostree sign` .

I've generated the key using
`openssl genpkey -algorithm ed25519 -out private.key`
and took only the base64-encoded part, and tried to pass it using both `--keys-file` and `<KEY-ID>` arguments to `ostree sign`. In either way, I get:
```
OT: using fuse: 0
OT: using fuse: 0
error: Invalid ed25519 secret key: Ill-formed input: expected 48 bytes, got 64 bytes
```
While inspecting the code, I found that the outputs of `validate_lenght` [function](https://github.com/ostreedev/ostree/blob/f911d40b986182d0d34b9c94d0d1b987da61d1ed/src/libostree/ostree-sign-ed25519.c#L109) are probably swapped, but also that the comments suggest that the API accepts also keys in binary format. So I've tried OpenSSL key in DER format, getting output:
```
OT: using fuse: 0
OT: using fuse: 0
error: Invalid ed25519 secret key: Ill-formed input: expected 0 bytes, got 64 bytes
```
Could you please take a look? 😅 I'm using latest Fedora Kinoite 41, the output of `ostree --version:`
```
libostree:
 Version: '2024.9'
 Git: 14e29775963d44bc533715efcf2362d6051756de
 Features:
  - inode64
  - initial-var
  - libcurl
  - libsoup3
  - gpgme
  - composefs
  - ex-fsverity
  - libarchive
  - selinux
  - openssl
  - sign-ed25519
  - libmount
  - systemd
  - release
  - p2p
```
Resources I've been following:
https://ostreedev.github.io/ostree/composefs/
https://blogs.gnome.org/alexl/2024/01/15/testing-composefs-in-silverblue/
#2867

**Comment by prydom** [comment, marked as answer]
You can use this script I wrote, that I mostly sourced from OSTree's tests. In particular this test:

https://github.com/ostreedev/ostree/blob/111a45fa5c3d115074d80b0e01f419969221fb63/tests/libtest.sh#L760-L776

For some reason the sign command wants the concatenated private key seed and the public key rather then just the private key seed like e.g. Wireguard. It also doesn't want the PKCS#8 header that OpenSSL generates.

```bash
#!/usr/bin/bash
umask=$(umask)

pemfile=private.pem
umask 177
openssl genpkey -algorithm ed25519 -outform PEM -out "${pemfile}"

# Extract the private and public parts from generated key.
ED25519PUBLIC="$(openssl pkey -outform DER -pubout -in ${pemfile} | tail -c 32 | base64)"
ED25519SEED="$(openssl pkey -outform DER -in ${pemfile} | tail -c 32 | base64)"
# Secret key is concantination of SEED and PUBLIC
ED25519SECRET="$(echo ${ED25519SEED}${ED25519PUBLIC} | base64 -d | base64 -w 0)"

echo $ED25519SEED > secret.ed25519
echo $ED25519SECRET > secret_concatinated.ed25519
umask $umask
echo $ED25519PUBLIC > public.ed25519
rm -f $pemfile

echo "Generated ed25519 keys:"
echo "public: ${ED25519PUBLIC}"

```

**Reply by ganjiii3** [reply]
Thank you! I should have looked at the tests 😄
